src/bin/genemichaels.rs

```diff
@@ -199,7 +237,104 @@ fn main() {
+                            workspace_cargo_toml = Some(d.join("Cargo.toml"));
+                            break;
+                        }
+                        at = d.parent();
+                    }
+                }
+                let (dirs, working_folder) = match workspace_cargo_toml {
+                    Some(ct) => {
+                        // if workspace exists
+                        let path = ct.canonicalize()?;
+                        let workspace_folder = path.parent().unwrap();
+                        let manifest = cargo_manifest::Manifest::from_path(&path)?;
+                        match manifest.workspace.map(|wkspc| wkspc.members) {
+                            Some(mut mb) => {
+                                mb.push(String::new());
+                                (mb, workspace_folder.to_path_buf())
+                            },
+                            // if no members, assume single folder workspace
+                            None => (Vec::from([String::new()]), c_dir),
+                        }
+                    },
+                    // if just checking current folder, same logic as single folder workspace
+                    None if args.folder => {
+                        if !c_dir.join("Cargo.toml").exists() {
+                            return Err(anyhow::anyhow!("No Cargo.toml found in current folder!"));
+                        }
+                        (Vec::from([String::new()]), c_dir)
+                    },
+                    None => {
+                        return Err(anyhow::anyhow!("No workspace found!"));
+                    },
+                };
+                let mut threads = Vec::new();
+                for dir in dirs {
+                    let mut path = working_folder.to_path_buf();
+                    path.push(dir);
+                    for f in walkdir::WalkDir::new(&path) {
+                        match f {
+                            Ok(file) => {
+                                let file_path = file.path().to_path_buf();
+                                if file_path.extension() == Some(OsStr::new("rs")) &&
+                                    !file_path.components().any(|c| c.as_os_str() == OsStr::new("target")) {
+                                    let args = args.clone();
+                                    threads.push(thread::spawn(move || {
+                                        process_workspace_file(&args, config, file.path().to_path_buf());
+                                    }));
+                                }
+                            },
+                            Err(e) => {
+                                eprintln!("Error opening file {}: {}", path.to_string_lossy(), e);
+                                continue;
+                            },
+                        }
+                    }
+                }
+                loop {
+                    if threads.iter().map(|j| j.is_finished()).all(|f| f) {
```

Comment:
You should be able to just iterate the thread list and `.join()` them, avoiding busy looping.  However, I wonder if it would be worth using something like https://docs.rs/threadpool/latest/threadpool/ which has a join method will set the max concurrency to something reasonable (host core count?) to avoid overwhelming the system.

Reply:
Good idea to save busy looping.

Reply:
Did you have any thoughts on using threadpool or another mechanism to limit the number of threads?

Reply:
Not particularly - having also tried using futures, and found `std::thread`s are faster, I'm not sure why we'd need to limit the amount of threads - is there a strong reasoning you have for not accessing all the available CPU resources while running?

Reply:
Since these are cpu bound, having more threads than cores will probably end up with them task switching and being overall less efficient (i.e. only one thread can be running on a core at once, and the os will be switching threads while they're running).

But also - doing 2 files in parallel will take 2x memory, and if there are large files that may end up with out of memory.

Reply:
Seems reasonable, implemented.